### ISSUE TITLE
"Did you mean...?" suggestions

### ISSUE THREAD

**Issue by brianpierson2020** [deluan/navidrome#255]
Maybe we could introduce a feature that asks "Did you mean... ?" showing the closes matches when there are not matches.

This could be incredible helpful on big libraries when making a typo of when the spelling is just a little different like US or UK English.

Another option (less fancy) might be to include some kind of fuzzy search and/or allowing wildcards.

**Comment by deluan** [comment]
This requires a proper Full Text Search implementation. Subsonic and its Java clones use the awesome Lucene library (that would support a "did you mean" feature easily), but it would be too heavyweight to use it in Navidrome. 

There are options in the Go-land (like [Bleve](https://github.com/blevesearch)), but they are usually slow and takes a lot of memory, going against of of the goals of Navidrome: to be as lightweigth as possible.

Having said that, I plan to move the search to SQLite's Full Text Search implementation. This would help with typos and more relevant search results, and maybe would allow a "Did you mean" like functionality in the future.

**Comment by SpencerDub** [comment]
> Having said that, I plan to move the search to SQLite's Full Text Search implementation.

@deluan, would that allow for a single universal search bar instead of the current view-specific search (search albums/tracks/artists)?

**Comment by deluan** [comment]
A single search bar is more of a UX concern than a technical implementation. The current "search" fields are actually filters for the current view. 

It is a different issue, feel free to create a new GH issue for it, although it would help having the SQLite FTS working first.

**Comment by orlea** [comment]
Other than English, for example Japanese are not separated by spaces per word, so they can't use the full-text search for English .

It would be hard to make the full-text search work for various languages.
So it would be very nice if partial match search could be used instead of full text search, depending on the setting.

For example, if you want to search for the word "音楽"
`LIKE '%音楽%'` in SQL

I expect it to be very slow, but I still think better results than putting Chinese, Japanese, Korean, etc. into a full-text search for English.

**Comment by orlea** [comment]
https://github.com/deluan/navidrome/compare/v0.31.0...orlea:feature/simple-search

This modified code actually works well enough in my 8000 song music library. However, it's probably bad for English etc.
I think it would be better to make it configurable with environment variables like `ND_SEARCHSEPARATOR(" " or "")`.
I don't know how to read and use settings from environment variables etc., as I'm a programming novice.

**Comment by deluan** [comment]
Hi @orlea , thanks for looking into this! I did some quick research for FTS in Asian languages and doesn't seem to exist a simple solution to the problem. I'll make the separator configurable as you suggested.

Whenever I implement a real (western) FTS in Navidrome, I'll keep the current implementation as a configurable option.

**Comment by orlea** [comment]
@deluan 
Thank you. I tried the `develop` tag docker image. (Version 0.31.0-SNAPSHOT (76e5227))
The search that was a problem for my use case has been solved!

I tried emby, airsonic, and navidrome in parallel, but I liked navidrome the best.
It's the lightest and fastest.

**Comment by tryffel** [comment]
Any update on this? 
I have personally used Meilisearch in multiple personal projects and I really love it for typo-tolerant search, for instance. 
I could try to create a POC / MR for integrating Meilisearch into Navidrome?

**Comment by avh4** [comment]
Just to add a concrete example from my actual library, I was searching songs today for "bird" and expected both "Blackbird" and "Bluebird" to appear in the results, but it currently doesn't because of the way search currently works.

**Comment by kaanchan** [comment]
> [Universal Search Bar] is a different issue, feel free to create a new GH issue for it, although it would help having the SQLite FTS working first.

@deluan What's not working, if anything, for SQLite FTS? Are you considering implementing the above?

**Comment by Szeraax** [comment]
IIRC, if you moved the backend from Sqlite to Postgres, you'd be able to get fuzzy matching (Cats matching Cat) without any work. Could be a consideration for something one day.

**Comment by mist2t** [comment]
A great add-on will be also the support of diacritics: a non-diacritic search will include the same word with diacritics in results and vice-versa

Example

Search for : Casa
Results: Casa, Cașa, Cășa

Etc.

**Comment by deluan** [comment]
This is already working:
![Screenshot 2024-11-20 at 6 54 40 AM](https://github.com/user-attachments/assets/373bdee1-f99e-4527-aea0-ef321db03e21)

Maybe not for this specific diacritic?

**Comment by mist2t** [comment]
> Maybe not for that specific diacritic?
> 
> EDIT: this is the full list of diacritics that are handled: https://github.com/deluan/sanitize/blob/master/sanitize.go#L218

Yeah, there are some missing diacritics in the list, the ones below are not present (both uppercase and lowercase)

Ț / ț
Ă / ă
Ș / ș

* found on Romanian localization / keyboard

**Comment by deluan** [comment]
@mist2t Fixed in https://github.com/navidrome/navidrome/commit/054946dc42304edbbc25b64f602da43998cced57

**Comment by Szeraax** [comment]
Will FTS allow users to search by all tags, such as "comment" and others?

**Comment by deluan** [comment]
Yep, FTS will search in all tags imported, but with different weights. Not sure what will be the UX for that, but this is a problem for when it is time to implement it.

**Comment by bilogic** [comment]
May I suggest a progressive approach of just listing the results first? and then work up the ladder, i.e. introducing weights etc

**Comment by deluan** [comment]
@Szeraax 
> IIRC, if you moved the backend from Sqlite to Postgres, you'd be able to get fuzzy matching (Cats matching Cat) without any work. Could be a consideration for something one day.

This is supported by Sqlite3 FTS as well: https://sqlite.org/fts5.html#porter_tokenizer

@bilogic 
The Sqlite FTS module already supports relevance. Given weights for each fields should not be that complicated, and we should not release the feature without it. Ex: when searching for "Yesterday", we would expect the Beatle's song to appear _before_ Mirrors by Justin Timberlake (that contains that word in the lyrics), right? For that to work, we need to give weights for certain fields

**Comment by bilogic** [comment]
In my experience of searching, I don't always search for the same field, e.g it could be the song's title, sometimes, I remember the band's name, sometimes it's the singer's name.

I mean, one of the main issues I had in Navidrome's search is not being able to figure out how it works easily. I clearly know it is in my collection, yet I can't find it, e.g. `ND_SEARCHFULLSTRING` and path, which I can see on the ND's UI, but not part of searched fields.

Perhaps a more contextual way is to list the results is by time of the record, and then after some use, add a popularity section (how many times the user has clicked on the record), record here referring to the database entry.

To further put things in context, I don't see this as a search engine like google, how many results do we expect when searching "yesterday" on a typical ND user's collection? maximum 100s? probably less than 50 to be more reasonable which can be fully combed through by the user in about 1-2 mins. I did a search on mine, 70 results, which I can humanly scan through, much better than 0 result, from which I can take no further action.

My point being, I'm not sure there is a universal set of weights that make sense for everyone. Weights probably make sense only when additional context is available, e.g. the person's age, geographic location, gender etc and it my case, it depends on my state of mind for that particular day.

I mean, if the weights are already designed, then sure, but otherwise, is it necessary to spend the effort?

**Comment by theacodes** [comment]
Have y'all considered Meilisearch? It's specifically focused on instant search stuff, which seems like a solid fit for something like navidrome. There's a Jellyfin plugin that uses it as well. 

Honestly it might make the most sense to make it possible to provide search engines via plugins, which could keep navidrome itself lightweight and sidestep having to incorporate an entire fts engine into the codebase.

More info: https://www.meilisearch.com/docs/learn/resources/comparison_to_alternatives

**Comment by kausar007** [comment]
+1 for Meilisearch. Another app that I use that uses Meilisearch is [Karakeep](https://github.com/karakeep-app/karakeep). The search within that app is awesome.

**Comment by deluan** [comment]
Meilisearch is a Rust server, not a Go library, so we can't have a hard dependency on external servers.

This will be implemented using SQLite FTS support.

**Comment by theacodes** [comment]
Also worth considering Typesense, as that is how Listenbrainz implements
its search. It's similar to Meilisearch - a separate program (C++ instead
of rust) which handles type ahead search.

Sqlite FTS is nice and would be a great improvement, but tools like
Meilisearch & Typesense are purpose built for this kind of search and would
be significantly faster.

Again, plugins would be great here. That would enable sqlite FTS,
Meilisearch, Typesense, and more without adding any hard dependencies or
requiring Navidrome to maintain the non-sqlite implementations.

These are just suggestions and explorations. No pressure, do this how ever
makes sense to you. 🙂

P.S.: I have some hacky Python that indexes songs in Meilisearch and uses
that to fuzzy match tracks for importing playlists, I can distill some of
that into something that can other folks can look at.

**Comment by deluan** [comment]
With the new plugin system in #4833 , it may be possible, but just keep in mind that while the search may be faster than SQLite's FTS implementation, indexing the music can significantly slow the scanner down.

To allow plugins to implement search, I'll need to restructure how search works currently, then introduce a new SearchProvider capability to the plugin system. I may spend some time with it before next release, let's see.

**Comment by theacodes** [comment]
Okay, here's something to demonstrate how absurdly fast Typesense and Meilisearch are. This is me doing a few queries in Navidrome:

https://github.com/user-attachments/assets/7df9ccb9-ce8a-4bd9-bbb5-58d6f018451b

& This is the same queries in Meilisearch:

https://github.com/user-attachments/assets/a2a31c26-4dd0-4f2a-98fb-c473f12fc2ce

Sqlite FTS will be more comprehensive but also slower than the current Navidrome search, and my personal feeling is that search speed is super important and Typesense/Meilisearch can give both.

> indexing the music can significantly slow the scanner down.

For sure, it probably makes sense to let plugins do it asynchronously. Though, it's less slow than you think- my library is 22,454 songs and Meilisearch happily slurps them up in about 2 seconds:

```
document indexing done ... { indexed_documents: 12450, number_of_documents: 12500 } processed_in=1.114082984s
document indexing done ... { indexed_documents: 9954, number_of_documents: 22454 } processed_in=885.48828ms
```

It's so fast that when I'm using this I literally just delete the index and re-create it every time, lol. It takes longer for my script to fetch and decode all of the data from Navidrome than it does to add all of the songs to Meilisearch.

This is on an `i7-8700 @ 4.60GHz` with `48G RAM`.

> To allow plugins to implement search, I'll need to restructure how search works currently, then introduce a new SearchProvider capability to the plugin system. I may spend some time with it before next release, let's see.

Rad, let me know if you do and perhaps I'll finally pick Go back up and write a plugin. :)

For future self/other folks who might wanna tackle this, this is roughly how I mapped song data into Meilisearch:

```python
dict(
    id=song.id,
    title=song.title,
    album=song.album,
    artist=song.artist,
    albumArtist=song.albumArtist,
    artists=list(
        set(
            p.name
            for p in (
                *song.participants.get("artist", []),
                *song.participants.get("albumartist", []),
            )
        )
    ),
    path=song.path,
    genre=list(
        set(
            filter(
                None,
                (song.genre, *[g.name for g in song.genres or []]),
            )
        )
    ),
    year=song.year,
    releaseYear=song.releaseYear,
    originalYear=song.originalYear,
    coverArtPreviewUrl=str(song.cover_art_url(size=300)),
    # Used to prefer songs on earlier discs, helpful for remasters/releases/deluxe albums with live/demo tracks
    discNumber=song.discNumber or 0,
    trackNumber=song.trackNumber or 0,
)
```

I have customized the index settings slightly, but really just adding some synonyms and tweaking the ranking to favor newer releases when all else is the same:

```json
{
    "searchableAttributes": [
        "title",
        "artist",
        "artists",
        "album",
        "albumArtist",
        "path",
        "genre",
        "year",
        "releaseYear",
        "originalYear"
    ],
    "filterableAttributes": [
        "genre",
        "year",
        "releaseYear",
        "originalYear"
    ],
    "sortableAttributes": [],
    "rankingRules": [
        "words",
        "typo",
        "proximity",
        "attribute",
        "sort",
        "exactness",
        "discNumber:asc",
        "trackNumber:asc"
    ],
    "stopWords": [],
    "nonSeparatorTokens": [],
    "separatorTokens": [],
    "dictionary": [],
    "synonyms": {
        "&": [
            "and"
        ],
        "and": [
            "&"
        ],
        "feat": [
            "featuring",
            "ft"
        ],
        "featuring": [
            "ft",
            "feat"
        ],
        "ft": [
            "featuring",
            "feat"
        ],
        "part": [
            "pt"
        ],
        "parts": [
            "pts"
        ],
        "point": [
            "pt"
        ],
        "points": [
            "pts"
        ],
        "pt": [
            "part",
            "point"
        ],
        "pts": [
            "parts",
            "points"
        ],
        "u": [
            "you"
        ]
    }
}
```

**Comment by deluan** [comment]
> Okay, here's something to demonstrate how absurdly fast Typesense and Meilisearch are. This is me doing a few queries in Navidrome:
> 
>  Screen.Recording.2025-12-31.at.4.44.39.PM.mov 
> & This is the same queries in Meilisearch:
> 
>  Screen.Recording.2025-12-31.at.4.44.16.PM.mov 

You don't have to tell me how efficient _any_ FTS implementation is :) Specially if we compare it ti the current search implementation that is just a simple scan on all rows, doing substring comparison.

> Sqlite FTS will be more comprehensive but also slower than the current Navidrome search

I'm pretty sure [SQLite FTS5](https://sqlite.org/fts5.html) will be really fast, and I dare to say it may be as fast as many other FTS implementation out there. The algorithms are the basically same, with few tweaks. What differentiate most of them is advanced features, like synonyms, facets, analyzers/tokenizers availability, handling misspelled words, language support, SDK, etc.. 

Having said that, I agree a dedicated piece of software could perform better than a plugin for a DB, but don't count on it being _extremely_ faster.


> and my personal feeling is that search speed is super important and Typesense/Meilisearch can give both.

Yes it is, but indexing time is also important in Navidrome, one of its key advantages over the competition

> 
> > indexing the music can significantly slow the scanner down.
> 
> For sure, it probably makes sense to let plugins do it asynchronously. Though, it's less slow than you think- my library is 22,454 songs and Meilisearch happily slurps them up in about 2 seconds:

Seems that you are accessing ND's DB directly, which won't be available for plugins. Maybe I'll separate the capabilities in Indexer (which could cause delays in ND) and Searcher. But users would have to provide their own solution to populate the search engine index, which is not great UX.

Again, let's see. Happy hollidays!

**Comment by theacodes** [comment]
> Seems that you are accessing ND's DB directly, which won't be available for plugins. 

Worse, I'm abusing the undocumented internal API.

> Maybe I'll separate the capabilities in Indexer (which could cause delays in ND) and Searcher.

Might make sense. I was thinking Navidrome would drive on a more granular level based on the scanner's outcome (e.g., `SearchProvider.updateSong()`, `removeSong()`, etc. instead of a big `SearchProvider.index()`), that way you could pass the song information into the plugin instead of the plugin having to ask the host for it.

**Comment by tryffel** [comment]
Also worth noting, there's the Go native `bleve` package that also has great support for typos/near matches. Could be worth trying it out. I have personally used sqlite's fts5 plugin and to me it seems like it doesn't really tackle the problems of similar results, which is what I personally miss the most. Fts5 queries are practically wildcard queries: when you query `beat` it returns `beat*` but not `*beat*` or better yet, `*bat*`, which is what you usually would want. But perhaps I'm just not using it right. Anyway, IMO Typesense/Meilisearch/Bleve user experience is just on whole other level compared to fts5 and the extra trouble is definitely worth it for the end user experience. 

Also, as mentioned earlier, I know that Meilisearch indexing is asynchronous, so the requests are fast and it takes some time (seconds to minutes, depending on queue size) until the documents actually appear in the search queries. 

Having used Meilisearch/Bleve & fts5, I'm happy to help implementing any of the aforementioned implementations.

**Comment by esiqveland** [comment]
+1 for a solution with Sqlite FTS to keep navidrome light and simple to run. It would be great to just get the search a little better than it currently is.

**Comment by deluan** [comment]
> +1 for a solution with Sqlite FTS to keep navidrome light and simple to run. It would be great to just get the search a little better than it currently is.

I agree. The current plan is:
- Implement an extensible Search interface for index/search.
- Implement the Search interface internally using SQLite FTS.
- Implement Plugin API to allow plugins to implement the Search interface. This would allow integration with any search engine.

**Comment by deluan** [comment]
For those that are following this issue, here's the FTS5 implementation: #5079 
Please give it a try (backup your DB first!!) and give feedback. 

I noticed based on the comments above that one thing that caused the impression of the search being slower than it actually is is the debounce that happens in the UI, that waits for the user to stop typing for 500ms before sending the search query. This is now configurable (`UISearchDebounceMs`) and the default was reduced to 200ms. You can try it with 0 to completely disable it, but keep in mind that each key you press will execute a search on the server.

~~This new implementation does not support the `SearchFullString` option (used mainly for CJK text), so if it is enabled Navidrome will automatically fallback to the legacy search implementation.~~ - CJK search is now supported automatically: a CJK search query will automatically fallback to the old SearchFullString mode.

Anyway, feedback is very welcome!
